internal/api/terminal_manager.go

```diff
@@ -0,0 +1,174 @@
+			Handler:     s.SendTerminalCommand,
+		},
+		{
+			RequestType: RequestTerminalResize,
+			Handler:     s.SendTerminalResize,
+		},
+	}
+}
+
+func (s *terminalStateManager) SendTerminalResize(state octant.State, payload action.Payload) error {
+	terminalID, err := payload.String("terminalID")
+	if err != nil {
+		return errors.Wrap(err, "extract terminal ID from payload")
+	}
+
+	rows, err := payload.Uint16("rows")
+	if err != nil {
+		return errors.Wrap(err, "extract rows from payload")
+	}
+
+	cols, err := payload.Uint16("cols")
+	if err != nil {
+		return errors.Wrap(err, "extract cols from payload")
+	}
+
+	tm := s.config.TerminalManager()
+	t, ok := tm.Get(terminalID)
+	if !ok {
+		return errors.New(fmt.Sprintf("terminal %s not found", terminalID))
+	}
+
+	if t.Active() {
+		t.Resize(cols, rows)
+	}
+	return nil
+}
+
+func (s *terminalStateManager) SendTerminalCommand(state octant.State, payload action.Payload) error {
+	terminalID, err := payload.String("terminalID")
+	if err != nil {
+		return errors.Wrap(err, "extract terminal ID from payload")
+	}
+
+	key, err := payload.String("key")
+	if err != nil {
+		return errors.Wrap(err, "extract key from payload")
+	}
+
+	tm := s.config.TerminalManager()
+	t, ok := tm.Get(terminalID)
+	if !ok {
+		return errors.New(fmt.Sprintf("terminal %s not found", terminalID))
+	}
+	return t.Write([]byte(key))
+}
+
+func (s *terminalStateManager) SendTerminalScrollback(state octant.State, payload action.Payload) error {
+	terminalID, err := payload.String("terminalID")
+	if err != nil {
+		return errors.Wrap(err, "extract terminal ID from payload")
+	}
+	s.setSendScrollback(terminalID, true)
+	return nil
+}
+
+func (s *terminalStateManager) setSendScrollback(id string, v bool) {
+	s.sendScrollback.Store(id, v)
+}
+
+func (s *terminalStateManager) Start(ctx context.Context, state octant.State, client OctantClient) {
+	ch := make(chan struct{}, 1)
+	defer func() {
+		close(ch)
+	}()
+	s.poller.Run(ctx, ch, s.runUpdate(state, client), event.TerminalStreamDelay)
+}
+
+func (s *terminalStateManager) runUpdate(state octant.State, client OctantClient) PollerFunc {
+	return func(ctx context.Context) bool {
+		tm := s.config.TerminalManager()
+		for _, t := range tm.List(state.GetNamespace()) {
+			line, err := t.Read(readBufferSize)
+			if err != nil {
+				t.SetExitMessage(fmt.Sprintf("%v\n", err))
+				t.Stop()
+				continue
+			}
+
+			sendScrollback, ok := s.sendScrollback.Load(t.ID())
+			if line == nil && (!ok || !sendScrollback.(bool)) {
+				continue
+			}
+
+			key := t.Key()
+			eventType := octant.EventType(fmt.Sprintf("terminals/namespace/%s/pod/%s/container/%s/%s", key.Namespace, key.Name, t.Container(), t.ID()))
+			data := terminalOutput{Line: line}
+
+			if ok && sendScrollback.(bool) {
+				data.Scrollback = t.Scrollback()
+				msg := t.ExitMessage()
+				if msg != "" {
+					data.Scrollback = append(data.Scrollback, []byte("\n"+msg)...)
+				}
+				s.setSendScrollback(t.ID(), false)
+			}
```

Comment:
The logic that decides if to send scrollback vs. line data is a bit hard to follow. There should just be a well defined interface that the TerminalManager calls to get a terminal event and then sends that event over the websocket.

This is leaking way too much detail in to the manager.